Feature/solved atm exercise

kerenGuez opened the pull request.

kerenGuez committed:
solved atm exercise.
my text atm file :
"""
322836790 pass1 2990
322893748 pass2 10987
322987654 pass3 -1000
322890987 pass4 567
"""

kerenGuez committed:
solved

kerenGuez committed:
shortened lines

kerenGuez committed:
shorter line

kerenGuez committed:
shortened line 157

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+
+
+def check_balance(customers_data, customer_id):
+    """Prints the given customer's balance.
```
imperative verbs in all functions

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+
+
+def get_a_number(message_input):
+    """ Gets a number from user and returns it.
```
Remove space at the start of the sentence

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+            print("This is not a number, try again.")
+            continue
+        break
+    return int(number)
```
pep 8

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+        customer_id (str): The specific customer id to work on.
+
+    Returns:
+        int: The given customer's balance.
```
the balance must be an int?

rfire01 replied:
?

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+    msg = "Enter the amount you want to withdrawal."
+    amount_to_withdrawal = get_a_number(msg)
+    curr_balance = get_balance(customers_data, customer_id)
+    new_balance = curr_balance - amount_to_withdrawal
```
someone can draw more money than he have in the bank?

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+    Args:
+        filename (str): The path to the atm details file.
+
+    Returns (dict): Dictionary of all customer's data from the file.
```
fix doc

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+        return customers
+    else:
+        print(f"the file path : '{filename}' doesn't exist")
+        return -1
```
pep 8

osherdp replied:
In Python we don't return `-1` to indicate a problem, we raise exceptions
Just let the `open` method raise the exception if the file does not exist

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+        return -1
+
+
+def switcher(customers_data, customer_id, choice):
```
give meaningful name for the function

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+    elif choice == "3":
+        cash_deposit(customers_data, customer_id)
+    else:
+        change_password(customers_data, customer_id)
```
pep8

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+            atm_file.write(line)
+
+
+def atm(filename):
```
this function is too long
a thumb rule - if the function is more then 15~20 lines, its probably do more than 1 logic, and it can be splited to smaller functions

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+        filename (str): The path to the file with the bank details.
+    """
+    data = get_data_from_file(filename)
+    if data != -1:
```
instead putting everything in indent, just check first and exit if needed

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+                    continue
+                switcher(data, customer_id, operation)
+        update_file(filename, data)
+
```
pep8

osherdp commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+            msg = "enter the ID of the customer you want (-1 to turn off).\n"
+            customer_id = input(msg)
+            if customer_id == "-1":
+                break
```
There are ways to make this loop simpler
Look at the lecture of Raymond (link in the repo) for how to do this

osherdp commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+    """
+    new_pass = input("Enter a new password.\n")
+    customers_data[customer_id][PASS_KEY] = new_pass
+    print(f"Password was changed to {new_pass}")
```
printing the password again is not a recommended action in my opinion

osherdp commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+
+
+def main():
+    path = r"C:\Users\keren\Desktop\atmDetails.txt"
```
It seems like only your computer can run this program, or at least all the Windows users that are named Keren
For this exercise and on (I saw it on other exercises as well), make it work with every computer

osherdp commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+                    print("you have to enter one of the options, try again.")
+                    continue
+                switcher(data, customer_id, operation)
+        update_file(filename, data)
```
It looks crowded
add a blank line after each block (if-block, for-block, etc.)

osherdp commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+                elif operation not in "1234":
+                    print("you have to enter one of the options, try again.")
+                    continue
+                switcher(data, customer_id, operation)
```
`switcher` is not an indicative name

osherdp commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,181 @@
+            customers = {}
+            for line in atm_file:
+                customer_id, the_pass, balance = line.split()
+                customer = {PASS_KEY: the_pass, BALANCE_KEY: balance}
```
why won't you cast `balance` right here to `int`?

kerenGuez committed:
fixed issues

kerenGuez committed:
fixed things

kerenGuez committed:
shorter line

kerenGuez committed:
shorter

kerenGuez committed:
space in between imports

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,198 @@
+
+ATM_FILE = 1
+PASS_KEY = "atm_password"
+BALANCE_KEY = "balance"
```
reorder

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,198 @@
+"""atm program which allows users to preform actions on customers."""
+from functools import partial
+
+import sys
```
fix imports ordering

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,198 @@
+    curr_balance = get_balance(customers_data, customer_id)
+    msg = "Enter the amount you want to withdrawal."
+    amount_to_withdrawal = get_a_number(msg)
+    if amount_to_withdrawal <= curr_balance:
+        new_balance = curr_balance - amount_to_withdrawal
+        customers_data[customer_id][BALANCE_KEY] = new_balance
+        print(f"operation was successful {customer_id} now has {new_balance}$")
```
blank line after logic block (here and everywhere else)

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,198 @@
+        atm(path)
+    except Exception as e:
+        print(e)
+        sys.exit()
```
same in the stam, if you do nothing with try except, just don't use it

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,198 @@
+def main():
+    """Get file path and call atm function"""
+    try:
+        path = sys.argv[ATM_FILE]
```
A little change to practice.
The bank manager want to see all clients balance, however, he don't know how to read the text file.
In order for the manager to read it, the data file should be in csv format.
Change your code to support csv data, instead of txt.

kerenGuez committed:
code now supports csv files, added new lines where needed, added my own explanations to exceptions. balance is not an int it is a float.

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,208 @@
+
+PASS_KEY = "atm_password"
+BALANCE_KEY = "balance"
+ATM_FILE = 1
```
rerorder

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,208 @@
+        cash_deposit(customers_data, customer_id)
+
+    else:
+        change_password(customers_data, customer_id)
```
try to think about more pythonic way to implement this

rfire01 commented on basic/atm/atm_exercise.py:
```diff
@@ -0,0 +1,208 @@
+        execute_choice(data, customer_id, operation)
+
+
+def atm(filename):
```
Its easy to steal money from people, you never ask from the user to enter his password to perform an action

kerenGuez committed:
Reordered imports and consts.
Implemented switch case differently.
Asks user for old pass in order to change new one

kerenGuez committed:
user has to enter password in order to perdorm any action on customer
